From 0cab74918f99057e8f203b1cffd7d1294e0d63db Mon Sep 17 00:00:00 2001
From: Jorge Ramirez-Ortiz <jorge@foundries.io>
Date: Fri, 6 Nov 2020 18:33:53 +0100
Subject: [PATCH] iMX8M: default make builds SPL

The flash.bin image therefore becomes the SPL image ready to be
signed.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>
---
 iMX8M/soc.mak          | 31 ++++++++++++++++++++++++
 iMX8M/u-boot.its.patch | 53 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 84 insertions(+)
 create mode 100644 iMX8M/u-boot.its.patch

diff --git a/iMX8M/soc.mak b/iMX8M/soc.mak
index b7b3986..6c2f1fb 100644
--- a/iMX8M/soc.mak
+++ b/iMX8M/soc.mak
@@ -88,6 +88,20 @@ lpddr4_dmem_1d = lpddr4_pmu_train_1d_dmem$(LPDDR_FW_VERSION).bin
 lpddr4_imem_2d = lpddr4_pmu_train_2d_imem$(LPDDR_FW_VERSION).bin
 lpddr4_dmem_2d = lpddr4_pmu_train_2d_dmem$(LPDDR_FW_VERSION).bin
 
+################################
+# spl.bin + dtb + pad + ddr_fw #
+################################
+u-boot-signed-spl-ddr.bin: u-boot-spl.bin $(lpddr4_imem_1d) $(lpddr4_dmem_1d) $(lpddr4_imem_2d) $(lpddr4_dmem_2d)
+	@objcopy -I binary -O binary --pad-to 0x8000 --gap-fill=0x0 $(lpddr4_imem_1d) lpddr4_pmu_train_1d_imem_pad.bin
+	@objcopy -I binary -O binary --pad-to 0x4000 --gap-fill=0x0 $(lpddr4_dmem_1d) lpddr4_pmu_train_1d_dmem_pad.bin
+	@objcopy -I binary -O binary --pad-to 0x8000 --gap-fill=0x0 $(lpddr4_imem_2d) lpddr4_pmu_train_2d_imem_pad.bin
+	@cat lpddr4_pmu_train_1d_imem_pad.bin lpddr4_pmu_train_1d_dmem_pad.bin > lpddr4_pmu_train_1d_fw.bin
+	@cat lpddr4_pmu_train_2d_imem_pad.bin $(lpddr4_dmem_2d) > lpddr4_pmu_train_2d_fw.bin
+	@cat u-boot-spl.bin u-boot-spl.dtb > u-boot-spl-dtb.bin
+	@dd if=u-boot-spl-dtb.bin of=u-boot-spl-pad.bin bs=4 conv=sync
+	@cat u-boot-spl-pad.bin lpddr4_pmu_train_1d_fw.bin lpddr4_pmu_train_2d_fw.bin > u-boot-signed-spl-ddr.bin
+	@rm -f u-boot-spl-pad.bin lpddr4_pmu_train_1d_fw.bin lpddr4_pmu_train_2d_fw.bin lpddr4_pmu_train_1d_imem_pad.bin lpddr4_pmu_train_1d_dmem_pad.bin lpddr4_pmu_train_2d_imem_pad.bin
+
 u-boot-spl-ddr.bin: u-boot-spl.bin $(lpddr4_imem_1d) $(lpddr4_dmem_1d) $(lpddr4_imem_2d) $(lpddr4_dmem_2d)
 	@objcopy -I binary -O binary --pad-to 0x8000 --gap-fill=0x0 $(lpddr4_imem_1d) lpddr4_pmu_train_1d_imem_pad.bin
 	@objcopy -I binary -O binary --pad-to 0x4000 --gap-fill=0x0 $(lpddr4_dmem_1d) lpddr4_pmu_train_1d_dmem_pad.bin
@@ -145,6 +159,15 @@ u-boot.itb: $(dtbs)
 	./mkimage_uboot -E -p 0x3000 -f u-boot.its u-boot.itb
 	@rm -f u-boot.its $(dtbs)
 
+u-boot-signed.itb: $(dtbs)
+	./$(PAD_IMAGE) tee.bin
+	./$(PAD_IMAGE) bl31.bin
+	./$(PAD_IMAGE) u-boot-nodtb.bin $(dtbs)
+	DEK_BLOB_LOAD_ADDR=$(DEK_BLOB_LOAD_ADDR) TEE_LOAD_ADDR=$(TEE_LOAD_ADDR) ATF_LOAD_ADDR=$(ATF_LOAD_ADDR) ./mkimage_fit_atf.sh $(dtbs) > u-boot.its
+	patch -u u-boot.its -i u-boot.its.patch
+	./mkimage_uboot -f u-boot.its u-boot.itb
+	./mkimage_uboot -F -k ${KEYDIR} -K ./u-boot-spl.dtb -r u-boot.itb
+
 dtbs_ddr3l = valddr3l.dtb
 $(dtbs_ddr3l):
 	./$(DTB_PREPROC) $(PLAT)-ddr3l-$(VAL_BOARD).dtb $(dtbs_ddr3l)
@@ -204,6 +227,8 @@ flash_ddr4_val: $(MKIMG) signed_hdmi_imx8m.bin u-boot-spl-ddr4.bin u-boot-ddr4.i
 else
 flash_evk: flash_evk_no_hdmi
 
+flash_spl: flash_evk_spl
+
 flash_evk_emmc_fastboot: flash_evk_no_hdmi_emmc_fastboot
 
 flash_ddr4_evk: flash_ddr4_evk_no_hdmi
@@ -214,6 +239,12 @@ flash_ddr4_val: flash_ddr4_val_no_hdmi
 
 endif
 
+##########################
+# Generate SPL bootable  #
+##########################
+flash_evk_spl: $(MKIMG) u-boot-signed.itb u-boot-signed-spl-ddr.bin
+	./mkimage_imx8 -version $(VERSION) -loader u-boot-signed-spl-ddr.bin $(SPL_LOAD_ADDR) -out $(OUTIMG)
+
 flash_evk_no_hdmi: $(MKIMG) u-boot-spl-ddr.bin u-boot.itb
 	./mkimage_imx8 -version $(VERSION) -fit -loader u-boot-spl-ddr.bin $(SPL_LOAD_ADDR) -second_loader u-boot.itb 0x40200000 0x60000 -out $(OUTIMG)
 
diff --git a/iMX8M/u-boot.its.patch b/iMX8M/u-boot.its.patch
new file mode 100644
index 0000000..3932ccc
--- /dev/null
+++ b/iMX8M/u-boot.its.patch
@@ -0,0 +1,53 @@
+--- file.a	2020-11-03 19:12:03.738888656 +0100
++++ file.b	2020-11-03 18:51:49.061424199 +0100
+@@ -11,12 +11,18 @@
+ 			arch = "arm64";
+ 			compression = "none";
+ 			load = <0x40200000>;
++			hash-1 {
++			       algo = "sha256";
++			};
+ 		};
+ 		fdt@1 {
+ 			description = "evk";
+ 			data = /incbin/("evk.dtb");
+ 			type = "flat_dt";
+ 			compression = "none";
++			hash-1 {
++			       algo = "sha256";
++			};
+ 		};
+ 		atf@1 {
+ 			description = "ARM Trusted Firmware";
+@@ -26,6 +32,9 @@
+ 			compression = "none";
+ 			load = <0x00920000>;
+ 			entry = <0x00920000>;
++			hash-1 {
++			       algo = "sha256";
++			};
+ 		};
+ 		tee@1 {
+ 			description = "TEE firmware";
+@@ -35,6 +44,9 @@
+ 			compression = "none";
+ 			load = <0xbe000000>;
+ 			entry = <0xbe000000>;
++			hash-1 {
++			       algo = "sha256";
++			};
+ 		};
+ 	};
+ 	configurations {
+@@ -45,6 +57,11 @@
+ 			firmware = "uboot@1";
+ 			loadables = "atf@1", "tee@1";
+ 			fdt = "fdt@1";
++                        signature {
++                                algo = "sha256,rsa2048";
++                                key-name-hint = "dev";
++                                sign-images = "firmware", "loadables", "fdt";
++                        };
+ 		};
+ 	};
+ };
-- 
2.17.1

